Reorg the monitoring task section (#32823)
kubernetes/website: content/en/docs/tasks/debug/debug-cluster/debug-cluster.md

---
reviewers:
- davidopp

> title: Troubleshoot Clusters

Reviewer: ```suggestion
title: Troubleshooting Clusters
```
For consistency?

description: Debugging common cluster issues.
content_type: concept
weight: 10

> ---

Reviewer 2: This would be good content for `content/en/docs/tasks/debug/debug-cluster/_index.md`

You could then link to the specific pages in the section within the page, and set `no_list: true` in the front matter to suppress the automatic section listing. I recommend you try it and see what you think.

<!-- overview -->

This doc is about cluster troubleshooting; we assume you have already ruled out your application as the root cause of the
problem you are experiencing. See
the [application troubleshooting guide](/docs/tasks/debug-application-cluster/applications/debug-application) for tips on application debugging.
You may also visit [troubleshooting document](/docs/tasks/debug-application-cluster/troubleshooting/) for more information.

<!-- body -->

## Listing your cluster

> The first thing to debug in your cluster is if your nodes are all registered correctly.
> 
> Run

Reviewer: ```suggestion
The first thing to debug in your cluster is if your nodes are all registered correctly.
Run the following command:
```

    status: "False"
    type: DiskPressure
  - lastHeartbeatTime: "2022-02-17T22:20:15Z"
    lastTransitionTime: "2022-02-17T22:13:25Z"
    message: kubelet has sufficient PID available
    reason: KubeletHasSufficientPID
    status: "False"
    type: PIDPressure
  - lastHeartbeatTime: "2022-02-17T22:20:15Z"
    lastTransitionTime: "2022-02-17T22:15:15Z"
    message: kubelet is posting ready status. AppArmor enabled
    reason: KubeletReady
    status: "True"
    type: Ready
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10250
  nodeInfo:
    architecture: amd64
    bootID: 22333234-7a6b-44d4-9ce1-67e31dc7e369
    containerRuntimeVersion: containerd://1.5.9
    kernelVersion: 5.13.0-28-generic
    kubeProxyVersion: v1.23.3
    kubeletVersion: v1.23.3
    machineID: 9384e2927f544209b5d7b67474bbf92b
    operatingSystem: linux
    osImage: Ubuntu 21.10
    systemUUID: aa829ca9-73d7-064d-9019-df07404ad448
```

## Looking at logs

For now, digging deeper into the cluster requires logging into the relevant machines.  Here are the locations
of the relevant log files.  (note that on systemd-based systems, you may need to use `journalctl` instead)

> ### Master

Reviewer: ```suggestion
### Control Plane nodes
```

> of the relevant log files.  (note that on systemd-based systems, you may need to use `journalctl` instead)

Reviewer 2: ```suggestion
of the relevant log files.  On systemd-based systems, you may need to use `journalctl` instead of
examining log files.
```

   * `/var/log/kube-apiserver.log` - API Server, responsible for serving the API
   * `/var/log/kube-scheduler.log` - Scheduler, responsible for making scheduling decisions

>    * `/var/log/kube-controller-manager.log` - Controller that manages replication controllers

Reviewer 2: ```suggestion
   * `/var/log/kube-controller-manager.log` - a component that runs most Kubernetes built-in {{<glossary_tooltip text="controllers" term_id="controller">}},
     with the notable exception of scheduling (the kube-scheduler handles scheduling).
```

However, you could just link to the relevant section of https://kubernetes.io/docs/concepts/overview/components/ and use a glossary tooltip for kube-apiserver, kube-scheduler, etc.
eg:
```markdown
* `/var/log/kube-apiserver.log` - {{< glossary_tooltip term_id="kube-apiserver" text="API server" >}} logs
```

### Worker Nodes

>    * `/var/log/kubelet.log` - Kubelet, responsible for running containers on the node
>    * `/var/log/kube-proxy.log` - Kube Proxy, responsible for service load balancing

Reviewer 2: ```suggestion
   * `/var/log/kubelet.log` - logs from the kubelet, responsible for running containers on the node
   * `/var/log/kube-proxy.log` - logs from `kube-proxy`, which is responsible for directing traffic to Service endpoints
```

> ## A general overview of cluster failure modes

Reviewer 2: ```suggestion
## Cluster failure modes
```

This is an incomplete list of things that could go wrong, and how to adjust your cluster setup to mitigate the problems.

> ### Root causes:

Reviewer: ```suggestion
### Root causes
```

> ### Root causes

Reviewer 2: ```suggestion
### Contributing causes
```
(I don't want to promote the ingle “root cause” concept)

  - VM(s) shutdown
  - Network partition within cluster, or between cluster and users
  - Crashes in Kubernetes software
  - Data loss or unavailability of persistent storage (e.g. GCE PD or AWS EBS volume)
  - Operator error, for example misconfigured Kubernetes software or application software

> ### Specific scenarios:

Reviewer: ```suggestion
### Specific scenarios
```

>   - Apiserver VM shutdown or apiserver crashing

Reviewer 2: nit: write “API server” not “Apiserver”.

    - Results
      - unable to stop, update, or start new pods, services, replication controller
      - existing pods and services should continue to work normally, unless they depend on the Kubernetes API
  - Apiserver backing storage lost
    - Results

>       - apiserver should fail to come up

Reviewer 2: ```suggestion
      - the kube-apiserver component fails to start successfully and become healthy
```

      - manual recovery or recreation of apiserver state necessary before apiserver is restarted
  - Supporting services (node controller, replication controller manager, scheduler, etc) VM shutdown or crashes
    - currently those are colocated with the apiserver, and their unavailability has similar consequences as apiserver
    - in future, these will be replicated as well and may not be co-located
    - they do not have their own persistent state
  - Individual node (VM or physical machine) shuts down
    - Results
      - pods on that Node stop running
  - Network partition
    - Results
      - partition A thinks the nodes in partition B are down; partition B thinks the apiserver is down. (Assuming the master VM ends up in partition A.)
  - Kubelet software fault
    - Results
      - crashing kubelet cannot start new pods on the node
      - kubelet might delete the pods or not
      - node marked unhealthy
      - replication controllers start new pods elsewhere
  - Cluster operator error
    - Results
      - loss of pods, services, etc
      - lost of apiserver backing store
      - users unable to read API
      - etc.

> ### Mitigations:

Reviewer: ```suggestion
### Mitigations
```
